<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
<meta name="pinterest" content="nopin">
<meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
<meta name="generator" content="Hugo 0.30" />



<link rel="canonical" href="/post/how-to-get-go-terms-from-refseq-ids/">


    <link href="//maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css" rel="stylesheet">
    <link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
    <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/8.4/styles/github.min.css">
    <title>How to get GO terms from Refseq IDs - 2-BitBio</title>
    
<meta name="description" content="A Gene Ontology analysis can add a lot of value to any omics study. Mapping GO terms to a newly sequenced genome or transcriptome can represent a challenge especially if the model system is… diverged. My typical functional annotation workflow usually involves.BLASTing gene sequences against RefSeq (although I typically use PLAST for this step since it’s much much fast than BLAST) BLASTing gene sequences against Uniprot databases Swissprot and Trembl.">

<meta property="og:title" content="How to get GO terms from Refseq IDs - 2-BitBio">
<meta property="og:type" content="article">
<meta property="og:url" content="/post/how-to-get-go-terms-from-refseq-ids/">
<meta property="og:image" content="/images/default.png">
<meta property="og:site_name" content="2-BitBio">
<meta property="og:description" content="A Gene Ontology analysis can add a lot of value to any omics study. Mapping GO terms to a newly sequenced genome or transcriptome can represent a challenge especially if the model system is… diverged. My typical functional annotation workflow usually involves.BLASTing gene sequences against RefSeq (although I typically use PLAST for this step since it’s much much fast than BLAST) BLASTing gene sequences against Uniprot databases Swissprot and Trembl.">
<meta property="og:locale" content="ja_JP">

<meta name="twitter:card" content="summary_large_image">
<meta name="twitter:site" content="2-BitBio">
<meta name="twitter:url" content="/post/how-to-get-go-terms-from-refseq-ids/">
<meta name="twitter:title" content="How to get GO terms from Refseq IDs - 2-BitBio">
<meta name="twitter:description" content="A Gene Ontology analysis can add a lot of value to any omics study. Mapping GO terms to a newly sequenced genome or transcriptome can represent a challenge especially if the model system is… diverged. My typical functional annotation workflow usually involves.BLASTing gene sequences against RefSeq (although I typically use PLAST for this step since it’s much much fast than BLAST) BLASTing gene sequences against Uniprot databases Swissprot and Trembl.">
<meta name="twitter:image" content="/images/default.png">


<script type="application/ld+json">
  {
    "@context": "http://schema.org",
    "@type": "NewsArticle",
    "mainEntityOfPage": {
      "@type": "WebPage",
      "@id":"/"
    },
    "headline": "How to get GO terms from Refseq IDs - 2-BitBio",
    "image": {
      "@type": "ImageObject",
      "url": "/images/default.png",
      "height": 800,
      "width": 800
    },
    "datePublished": "2020-01-18T00:00:00JST",
    "dateModified": "2020-01-18T00:00:00JST",
    "author": {
      "@type": "Person",
      "name": "2-BitBio"
    },
    "publisher": {
      "@type": "Organization",
      "name": "2-BitBio",
      "logo": {
        "@type": "ImageObject",
        "url": "/images/logo.png",
        "width": 600,
        "height": 60
      }
    },
    "description": "A Gene Ontology analysis can add a lot of value to any omics study. Mapping GO terms to a newly sequenced genome or transcriptome can represent a challenge especially if the model system is… diverged. My typical functional annotation workflow usually involves.
BLASTing gene sequences against RefSeq (although I typically use PLAST for this step since it’s much much fast than BLAST)
 BLASTing gene sequences against Uniprot databases Swissprot and Trembl."
  }
</script>


    <link href="/css/styles.css" rel="stylesheet">
    
    
    
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-111736277-3"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-111736277-3');
</script>


  </head>

  <body>
    
    
    

    <header class="l-header">
      <nav class="navbar navbar-default">
        <div class="container">
          <div class="navbar-header">
            <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
              <span class="sr-only">Toggle navigation</span>
              <span class="icon-bar"></span>
              <span class="icon-bar"></span>
              <span class="icon-bar"></span>
            </button>
            <a class="navbar-brand" href="/">2-BitBio</a>
          </div>

          

        </div>
      </nav>
    </header>

    <main>
      <div class="container">
        
<div class="row">
  <div class="col-md-8">

    <nav class="p-crumb">
      <ol class="breadcrumb">
        <li><a href="/"><i class="fa fa-home" aria-hidden="true"></i></a></li>
        
        <li itemscope="" itemtype="http://data-vocabulary.org/Breadcrumb"><a href="/post/" itemprop="url"><span itemprop="title">post</span></a></li>
        
        <li class="active">How to get GO terms from Refseq IDs</li>
      </ol>
    </nav>

    <article class="single">
  <header>
    <ul class="p-facts">
      <li><i class="fa fa-calendar" aria-hidden="true"></i><time datetime="2020-01-18T00:00:00JST">Jan 18, 2020</time></li>
      <li><i class="fa fa-bookmark" aria-hidden="true"></i><a href="/post/">post</a></li>
      
    </ul>

    <h1 class="title">How to get GO terms from Refseq IDs</h1>
  </header>

  

  <div class="article-body"><p><img src="../../img/mushroom-guardians.jpg" width="100%" /></p>
<p>A Gene Ontology analysis can add a lot of value to any omics study. Mapping GO terms to a newly sequenced genome or transcriptome can represent a challenge especially if the model system is… diverged. My typical functional annotation workflow usually involves.</p>
<ol style="list-style-type: decimal">
<li>BLASTing gene sequences against RefSeq (although I typically use PLAST for this step since it’s <a href="https://2-bitbio.com/2017/07/blastx-is-too-slow-heres-some.html">much much fast than BLAST</a>)<br />
</li>
<li>BLASTing gene sequences against Uniprot databases Swissprot and Trembl.<br />
</li>
<li>Retrieving GOterms using the Uniprot identifiers.</li>
</ol>
<p>It is possible however to skip the Uniprot BLASTs and retrieve the GO terms from Uniprot using the RefSeq IDs directly. I should note that this only works for gene models that exist in BOTH Refseq AND Uniprot and have 100% identity. So if you’re working on a non-traditional model system your mileage may vary.</p>
<p>You query Uniprot directly using their API with RefSeq as the search term. The following syntax will retrieve the GO terms (if any) for the <code>REFSEQID</code> term:<br />
<code>https://www.uniprot.org/uniprot/?query=database%3A%28type%3Arefseq+REFSEQID%29&amp;sort=score&amp;columns=id,entry%20name,go-id,go,go(biological%20process),go(molecular%20function),go(cellular%20component)&amp;format=tab</code></p>
<p>So if you have a BLAST tab file (<code>-outfmt 6</code>) from RefSeq it is fairly easy to loop all of this. The Python script below does just that. Simply feed it a file of RefSeq IDs and it will spit out the GO terms of all associated Uniprot IDs.</p>
<pre class="python"><code>import sys, os
from urllib.request import urlopen

##
## refseq2go.py
## By: Jake Warner
##

## This script takes a list of Refseq with IDs and outputs a tab deliminated file with the following fields:

# Refseq ID
# Uniprot ID
# Uniprot description
# GO IDs
# GO Biological process
# GO Molecular function
# GO Cellular Compartment

infile = open(sys.argv[1],&#39;r&#39;)
outfile = open(&#39;%s_go.txt&#39; %(sys.argv[1][:-4]), &#39;w&#39;)

# isolate refseq ID
linecount=0
uprothits=0
uprotmissing=0
#loop the query IDs
for line in infile:
    linecount+=1
    if linecount==1:
        outfile.write(&#39;refseq_id\tuniprot_id\tuniprot_description\tGO_ids\tGO_BP\tGO_MF\tGO_CC\n&#39;)
    line=line.rstrip()
    refseq_id=line
    page = urlopen(&#39;https://www.uniprot.org/uniprot/?query=database%3A%28type%3Arefseq+&#39;+refseq_id+&#39;%29&amp;sort=score&amp;columns=id,entry%20name,go-id,go,go(biological%20process),go(molecular%20function),go(cellular%20component)&amp;format=tab&#39;).read()
    try:
        page = page.decode(&#39;utf-8&#39;).splitlines()[1]
        uprothits+=1
    except:
        uprotmissing+=1
    try:
        uprot_id =page.split(&#39;\t&#39;)[0]
    except:
        uprot_id =&#39;No_Uniprot_ID&#39;
    try:
        uprot_description=page.split(&#39;\t&#39;)[1]
    except:
        uprot_description=&#39;No_Uniprot_description&#39;
    try:
        go_id=page.split(&#39;\t&#39;)[2]
    except:
        go_id=&#39;No_GO_Codes&#39;
    try:
        go_bp=page.split(&#39;\t&#39;)[4]
    except:
        go_bp=&#39;No_GO_codes&#39;
    try:
        go_mf=page.split(&#39;\t&#39;)[5]
    except:
        go_mf=&#39;No_GO_codes&#39;
    try:
        go_cc=page.split(&#39;\t&#39;)[6]
    except:
        go_cc=&#39;No_GO_codes&#39;
    outfile.write(&#39;%s\t%s\t%s\t%s\t%s\t%s\t%s\n&#39;%(refseq_id,uprot_id,uprot_description,go_id,go_bp,go_mf,go_cc))
print(&quot;Read &quot;+str(linecount)+&quot; Refseq IDs&quot;)
print(str(uprothits) +&quot; had a match in Uniprot&quot;)
print(str(uprotmissing) + &quot; had no Uniprot match&quot;)
outfile.close()</code></pre>
<p>You can invoke it like so:</p>
<pre class="bash"><code>python refseq2go.py refseqIDs.txt</code></pre>
<p>Happy annotating!</p>
</div>

  <footer class="article-footer">
    
    
    
    <section class="bordered">
      <header>
        <div class="panel-title">CATEGORIES</div>
      </header>
      <div>
        <ul class="p-terms">
          
          <li><a href="/categories/bioinformatics/">BioInformatics</a></li>
          
        </ul>
      </div>
    </section>
    
    
    
    <section class="bordered">
      <header>
        <div class="panel-title">TAGS</div>
      </header>
      <div>
        <ul class="p-terms">
          
          <li><a href="/tags/python/">Python</a></li>
          
        </ul>
      </div>
    </section>
    
    
  </footer>

</article>


    
  </div>

  <div class="col-md-4">
    
<aside class="l-sidebar">

  <section class="panel panel-default">
    <div class="panel-heading">
      <div class="panel-title">LATESTS</div>
    </div>
    <div class="list-group">
      
      <a href="/post/how-to-get-go-terms-from-refseq-ids/" class="list-group-item">How to get GO terms from Refseq IDs</a>
      
      <a href="/post/fuzzy-cmeans-clustering-of-rnaseq-data-using-mfuzz/" class="list-group-item">Fuzzy cMeans clustering of RNAseq data using mFuzz</a>
      
      <a href="/post/clustering-rnaseq-data-making-heatmaps-and-tree-cutting-to-identify-gene-modules/" class="list-group-item">Clustering RNAseq data, making heatmaps, and tree cutting to identify gene modules.</a>
      
      <a href="/post/clustering-rnaseq-data-using-k-means-how-many-clusters/" class="list-group-item">Clustering RNAseq data using K-means: how many clusters?</a>
      
      <a href="/post/one-hot-encode-a-dna-sequence-using-python-and-scikit-learn/" class="list-group-item">One hot encode a DNA sequence using python and scikit learn</a>
      
      <a href="/post/clustering-rnaseq-data-using-fuzzy-c-means-clustering/" class="list-group-item">Clustering RNAseq data using fuzzy c-means clustering</a>
      
    </div>
  </section>

  
  <section class="panel panel-default">
    <div class="panel-heading">
      <div class="panel-title">CATEGORY</div>
    </div>
    <div class="list-group">
      
      <a href="/categories/bioinformatics" class="list-group-item">bioinformatics</a>
      
    </div>
  </section>
  
  <section class="panel panel-default">
    <div class="panel-heading">
      <div class="panel-title">TAG</div>
    </div>
    <div class="list-group">
      
      <a href="/tags/clustering" class="list-group-item">clustering</a>
      
      <a href="/tags/r" class="list-group-item">r</a>
      
      <a href="/tags/rnaseq" class="list-group-item">rnaseq</a>
      
      <a href="/tags/machine-learning" class="list-group-item">machine-learning</a>
      
      <a href="/tags/python" class="list-group-item">python</a>
      
    </div>
  </section>
  

  <section class="panel panel-default">
    <div class="panel-heading">
      <div class="panel-title">About</div>
    </div>
    <div class="list-group">
	  <div class="list-group-item">
	    <p>2-BitBio is <a href="http://jacobwarner.net", target="none">Jake Warner</a></p>
	    <p><i style="font-size:24px" class="fa">&#xf099;</i><a href="https://twitter.com/ScientistJake", target="none">@ScientistJake</a></p>
	  </div>
    </div>
  </section>
  
  <section class="panel panel-default">
    <div class="panel-heading">
      <div class="panel-title">Popular Posts</div>
    </div>
    <div class="list-group">
	  <div class="list-group-item">
	    <p><a href="http://2-bitbio.com/2017/10/clustering-rnaseq-data-using-k-means.html", target="none">Clustering RNAseq data using K-means.</a> ;</p>
	    <p><a href="http://2-bitbio.com/2017/04/clustering-rnaseq-data-making-heatmaps.html", target="none">Clustering RNAseq data and tree cutting to identify gene modules.</a> ;</p>
	    <p><a href="http://2-bitbio.com/2017/04/animated-3d-pca-plots-in-r.html", target="none">Animated 3D PCA plots in R.</a> ;</p>
	    <p><a href="http://2-bitbio.com/2017/06/running-blast-in-shiny-web-application.html", target="none">Running BLAST in Shiny.</a> ;</p>
	    <p><a href="http://2-bitbio.com/2018/06/one-hot-encode-dna-sequence-using.html", target="none">One hot encoding DNA sequences using python.</a> ;</p>
	  </div>
    </div>
  </section>
    <section class="panel panel-default">
    <div class="panel-heading">
      <div class="panel-title">Archive</div>
    </div>
    <div class="list-group">
      <div class="list-group-item">
        <button class="accordion list-group-item">2018</button>
        <div class="accordian-panel panel">
         <p><a href="http://2-bitbio.com/2018/06/one-hot-encode-dna-sequence-using.html", target="none">One hot encoding DNA sequences using python.</a> ;</p>
          <p><a href="http://2-bitbio.com/2018/01/making-interactive-plots-using.html", target="none">Making interactive plots using javascript and D3 (v4)</a> ;</p>
        </div>
      </div>
    </div>
    <div class="list-group">
      <div class="list-group-item">
        <button class="accordion list-group-item">2017</button>
        <div class="accordian-panel">
          <p><a href="http://2-bitbio.com/2017/12/clickable-volcano-plots-in-shiny.html", target="none">Clickable Volcano Plots in Shiny!</a> ;</p>
         <p><a href="http://2-bitbio.com/2017/12/xkcd-plots.html", target="none">Xkcd plots!</a> ;</p>
         <p><a href="http://2-bitbio.com/2017/10/clustering-rnaseq-data-using-k-means.html", target="none">Clustering RNAseq data using K-means: how many clusters?</a> ;</p>
         <p><a href="http://2-bitbio.com/2017/09/bulk-download-stock-data-from-yahoo.html", target="none">Bulk download stock data from Yahoo finance with R</a> ;</p>
         <p><a href="http://2-bitbio.com/2017/09/pokemods-educational-outreach-initiative.html", target="none">Pokemodels: An outreach initiative</a> ;</p>
         <p><a href="http://2-bitbio.com/2017/07/blastx-is-too-slow-heres-some.html", target="none">BLASTx is too slow! Here's some alternatives.</a> ;</p>
         <p><a href="http://2-bitbio.com/2017/06/ramazon-get-your-shiny-app-on-amazon.html", target="none">Ramazon: Deploying shiny apps on AWS</a> ;</p>
         <p><a href="http://2-bitbio.com/2017/06/running-blast-in-shiny-web-application.html", target="none">Shiny BLAST</a> ;</p>
         <p><a href="http://2-bitbio.com/2017/05/using-shiny-to-make-web-plotter-for.html", target="none">Using Shiny to make a web plotter for RNAseq data</a> ;</p>
         <p><a href="http://2-bitbio.com/2017/04/animated-3d-pca-plots-in-r.html", target="none">Animated 3D PCA plots</a> ;</p>
         <p><a href="http://2-bitbio.com/2017/04/bash-commands-for-manipulating-fasta.html", target="none">Bash commands for manipulating fasta files</a> ;</p>
          <p><a href="http://2-bitbio.com/2017/04/clustering-rnaseq-data-making-heatmaps.html", target="none">Clustering RNAseq data, making heatmaps, and tree cutting to identify gene modules.</a> ;</p>
          <p><a href="http://2-bitbio.com/2017/04/rin-value-how-low-is-too-low.html", target="none">RIN value, how low is too low?</a> ;</p>
          <p><a href="http://2-bitbio.com/2017/03/rna-seq-do-i-need-replicates.html", target="none">RNAseq, how many replicates?</a> ;</p>
        </div>
      </div>
</aside>


  </div>
</div>

      </div>
    </main>

    <footer class="l-footer">
      <div class="container">
        <p><span class="h-logo">&copy; 2-BitBio</span></p>
        <aside>
          <p>Powered by <a href="https://gohugo.io/">Hugo</a>.</p>
          <p>Theme: <a href="https://github.com/dim0627/hugo_theme_beg">Beg</a></p>
        </aside>
      </div>
    </footer>

    <script src="//code.jquery.com/jquery-3.1.1.min.js"></script>
    <script src="//maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>
    <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/8.4/highlight.min.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>
    <script>
      var acc = document.getElementsByClassName("accordion");
      var i;

      for (i = 0; i < acc.length; i++) {
       acc[i].addEventListener("click", function() {
          

          this.classList.toggle("active");

          
         var panel = this.nextElementSibling;
         if (panel.style.display === "block") {
           panel.style.display = "none";
          } else {
            panel.style.display = "block";
         }
        });
      } 
    </script>
  </body>
</html>

