<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
<meta name="pinterest" content="nopin">
<meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
<meta name="generator" content="Hugo 0.30" />

<link rel="alternate" type="application/rss+xml" title="RSS" href="/categories/bioinformatics/index.xml">

<link rel="canonical" href="/categories/bioinformatics/">


    <link href="//maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css" rel="stylesheet">
    <link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
    <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/8.4/styles/github.min.css">
    <title>Bioinformatics - 2-BitBio</title>
    
    <link href="/css/styles.css" rel="stylesheet">
    
    
    
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-111736277-3"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-111736277-3');
</script>


  </head>

  <body>
    
    
    

    <header class="l-header">
      <nav class="navbar navbar-default">
        <div class="container">
          <div class="navbar-header">
            <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
              <span class="sr-only">Toggle navigation</span>
              <span class="icon-bar"></span>
              <span class="icon-bar"></span>
              <span class="icon-bar"></span>
            </button>
            <a class="navbar-brand" href="/">2-BitBio</a>
          </div>

          

        </div>
      </nav>
    </header>

    <main>
      <div class="container">
        
<div class="row">
  <div class="col-md-8">

    
    <header class="page-header">
      <h1>Bioinformatics</h1>
    </header>
    

    <ul class="p-articles">
      
      <li><article class="li">
  <header>
    <ul class="p-facts">
      <li><i class="fa fa-calendar" aria-hidden="true"></i><time datetime="2020-01-18T00:00:00JST">Jan 18, 2020</time></li>
      <li><i class="fa fa-bookmark" aria-hidden="true"></i><a href="/post/">post</a></li>
      
    </ul>
    <h2 class="title"><a href="/post/how-to-get-go-terms-from-refseq-ids/">How to get GO terms from Refseq IDs</a></h2>
  </header>

  
  <div class="summary">A Gene Ontology analysis can add a lot of value to any omics study. Mapping GO terms to a newly sequenced genome or transcriptome can represent a challenge especially if the model system is… diverged. My typical functional annotation workflow usually involves.
BLASTing gene sequences against RefSeq (although I typically use PLAST for this step since it’s much much fast than BLAST)
 BLASTing gene sequences against Uniprot databases Swissprot and Trembl.</div>

  
  <footer>
    <a href="/post/how-to-get-go-terms-from-refseq-ids/" title="How to get GO terms from Refseq IDs">Read More…</a>
  </footer>
  
</article>
</li>
      
      <li><article class="li">
  <header>
    <ul class="p-facts">
      <li><i class="fa fa-calendar" aria-hidden="true"></i><time datetime="2019-05-09T00:00:00JST">May 9, 2019</time></li>
      <li><i class="fa fa-bookmark" aria-hidden="true"></i><a href="/post/">post</a></li>
      
    </ul>
    <h2 class="title"><a href="/post/fuzzy-cmeans-clustering-of-rnaseq-data-using-mfuzz/">Fuzzy cMeans clustering of RNAseq data using mFuzz</a></h2>
  </header>

  
  <div class="summary">Note this is part 4 of a series on clustering RNAseq data. Check out part one on hierarcical clustering here ; part two on K-means clustering here ; and part three on fuzzy c-means clustering here.
Clustering is a useful data reduction technique for RNAseq experiments. In previous posts, we discussed the usefulness of hard clustering techniques such as hierarcical clustering and K-means clustering. These techniques will partition all genes into co-expression clusters.</div>

  
  <footer>
    <a href="/post/fuzzy-cmeans-clustering-of-rnaseq-data-using-mfuzz/" title="Fuzzy cMeans clustering of RNAseq data using mFuzz">Read More…</a>
  </footer>
  
</article>
</li>
      
      <li><article class="li">
  <header>
    <ul class="p-facts">
      <li><i class="fa fa-calendar" aria-hidden="true"></i><time datetime="2019-05-08T00:00:00JST">May 8, 2019</time></li>
      <li><i class="fa fa-bookmark" aria-hidden="true"></i><a href="/post/">post</a></li>
      
    </ul>
    <h2 class="title"><a href="/post/clustering-rnaseq-data-making-heatmaps-and-tree-cutting-to-identify-gene-modules/">Clustering RNAseq data, making heatmaps, and tree cutting to identify gene modules.</a></h2>
  </header>

  
  <div class="summary">From the archive (originally published 2017-04-04): Clustering is extremely useful for generating hypotheses and data exploration in general. The idea is that genes which have similar expression patterns (co-expression genes) are often controlled by the same regulatory mechanisms (co-regulated genes). Often times co-expressed genes share similar functions so by looking at which genes are found in a cluster we can get an idea of what that cluster is doing. Here we’ll show how to cluster RNAseq data using hierarchical clustering.</div>

  
  <footer>
    <a href="/post/clustering-rnaseq-data-making-heatmaps-and-tree-cutting-to-identify-gene-modules/" title="Clustering RNAseq data, making heatmaps, and tree cutting to identify gene modules.">Read More…</a>
  </footer>
  
</article>
</li>
      
      <li><article class="li">
  <header>
    <ul class="p-facts">
      <li><i class="fa fa-calendar" aria-hidden="true"></i><time datetime="2019-02-04T00:00:00JST">Feb 4, 2019</time></li>
      <li><i class="fa fa-bookmark" aria-hidden="true"></i><a href="/post/">post</a></li>
      
    </ul>
    <h2 class="title"><a href="/post/clustering-rnaseq-data-using-k-means-how-many-clusters/">Clustering RNAseq data using K-means: how many clusters?</a></h2>
  </header>

  
  <div class="summary">From the archive: Clustering gene expression data allows us to identify substructures in the data and identify groups of genes that behave similarly. This method can help us identify genes that share a biological function (co-functional) and genes that are under the same control logic (co-regulated). Here we’ll show how to cluster RNAseq data using K-means clustering. We’ll address picking the appropriate clustering number then we’ll test drive some visualizations and plots.</div>

  
  <footer>
    <a href="/post/clustering-rnaseq-data-using-k-means-how-many-clusters/" title="Clustering RNAseq data using K-means: how many clusters?">Read More…</a>
  </footer>
  
</article>
</li>
      
      <li><article class="li">
  <header>
    <ul class="p-facts">
      <li><i class="fa fa-calendar" aria-hidden="true"></i><time datetime="2018-09-04T00:00:00JST">Sep 4, 2018</time></li>
      <li><i class="fa fa-bookmark" aria-hidden="true"></i><a href="/post/">post</a></li>
      
    </ul>
    <h2 class="title"><a href="/post/one-hot-encode-a-dna-sequence-using-python-and-scikit-learn/">One hot encode a DNA sequence using python and scikit learn</a></h2>
  </header>

  
  <div class="summary">From the archive: Machine learning (in the informatics world) is like teenage sex: everyone talks about it, nobody really knows how to to do it, everyone thinks everyone else is doing it, so everyone claims they are too. Juvenile comparisons aside, the power of these tools can’t be ignored. Before applying most machine learning algorithms to DNA sequences they must first be converted to binary strings. Here we’ll show how to one hot encode a DNA sequence in Python using SciKit Learn.</div>

  
  <footer>
    <a href="/post/one-hot-encode-a-dna-sequence-using-python-and-scikit-learn/" title="One hot encode a DNA sequence using python and scikit learn">Read More…</a>
  </footer>
  
</article>
</li>
      
      <li><article class="li">
  <header>
    <ul class="p-facts">
      <li><i class="fa fa-calendar" aria-hidden="true"></i><time datetime="2018-07-31T00:00:00JST">Jul 31, 2018</time></li>
      <li><i class="fa fa-bookmark" aria-hidden="true"></i><a href="/post/">post</a></li>
      
    </ul>
    <h2 class="title"><a href="/post/clustering-rnaseq-data-using-fuzzy-c-means-clustering/">Clustering RNAseq data using fuzzy c-means clustering</a></h2>
  </header>

  
  <div class="summary">Note this is part 3 of a series on clustering RNAseq data. Check out part one on hierarcical clustering here and part two on K-means clustering here.
Clustering gene expression is a particularly useful data reduction technique for RNAseq experiments. It allows us to bin genes by expression profile, correlate those bins to external factors like phenotype, and discover groups of co-regulated genes. Two common methods for clustering are hierarchical (agglomerative) clustering and k-means (centroid based) clustering which we discussed in part one and part two of this series.</div>

  
  <footer>
    <a href="/post/clustering-rnaseq-data-using-fuzzy-c-means-clustering/" title="Clustering RNAseq data using fuzzy c-means clustering">Read More…</a>
  </footer>
  
</article>
</li>
      
    </ul>

    


  </div>
  <div class="col-md-4">
    
<aside class="l-sidebar">

  <section class="panel panel-default">
    <div class="panel-heading">
      <div class="panel-title">LATESTS</div>
    </div>
    <div class="list-group">
      
      <a href="/post/how-to-get-go-terms-from-refseq-ids/" class="list-group-item">How to get GO terms from Refseq IDs</a>
      
      <a href="/post/fuzzy-cmeans-clustering-of-rnaseq-data-using-mfuzz/" class="list-group-item">Fuzzy cMeans clustering of RNAseq data using mFuzz</a>
      
      <a href="/post/clustering-rnaseq-data-making-heatmaps-and-tree-cutting-to-identify-gene-modules/" class="list-group-item">Clustering RNAseq data, making heatmaps, and tree cutting to identify gene modules.</a>
      
      <a href="/post/clustering-rnaseq-data-using-k-means-how-many-clusters/" class="list-group-item">Clustering RNAseq data using K-means: how many clusters?</a>
      
      <a href="/post/one-hot-encode-a-dna-sequence-using-python-and-scikit-learn/" class="list-group-item">One hot encode a DNA sequence using python and scikit learn</a>
      
      <a href="/post/clustering-rnaseq-data-using-fuzzy-c-means-clustering/" class="list-group-item">Clustering RNAseq data using fuzzy c-means clustering</a>
      
    </div>
  </section>

  
  <section class="panel panel-default">
    <div class="panel-heading">
      <div class="panel-title">CATEGORY</div>
    </div>
    <div class="list-group">
      
      <a href="/categories/bioinformatics" class="list-group-item">bioinformatics</a>
      
    </div>
  </section>
  
  <section class="panel panel-default">
    <div class="panel-heading">
      <div class="panel-title">TAG</div>
    </div>
    <div class="list-group">
      
      <a href="/tags/clustering" class="list-group-item">clustering</a>
      
      <a href="/tags/r" class="list-group-item">r</a>
      
      <a href="/tags/rnaseq" class="list-group-item">rnaseq</a>
      
      <a href="/tags/machine-learning" class="list-group-item">machine-learning</a>
      
      <a href="/tags/python" class="list-group-item">python</a>
      
    </div>
  </section>
  

  <section class="panel panel-default">
    <div class="panel-heading">
      <div class="panel-title">About</div>
    </div>
    <div class="list-group">
	  <div class="list-group-item">
	    <p>2-BitBio is <a href="http://jacobwarner.net", target="none">Jake Warner</a></p>
	    <p><i style="font-size:24px" class="fa">&#xf099;</i><a href="https://twitter.com/ScientistJake", target="none">@ScientistJake</a></p>
	  </div>
    </div>
  </section>
  
  <section class="panel panel-default">
    <div class="panel-heading">
      <div class="panel-title">Popular Posts</div>
    </div>
    <div class="list-group">
	  <div class="list-group-item">
	    <p><a href="http://2-bitbio.com/2017/10/clustering-rnaseq-data-using-k-means.html", target="none">Clustering RNAseq data using K-means.</a> ;</p>
	    <p><a href="http://2-bitbio.com/2017/04/clustering-rnaseq-data-making-heatmaps.html", target="none">Clustering RNAseq data and tree cutting to identify gene modules.</a> ;</p>
	    <p><a href="http://2-bitbio.com/2017/04/animated-3d-pca-plots-in-r.html", target="none">Animated 3D PCA plots in R.</a> ;</p>
	    <p><a href="http://2-bitbio.com/2017/06/running-blast-in-shiny-web-application.html", target="none">Running BLAST in Shiny.</a> ;</p>
	    <p><a href="http://2-bitbio.com/2018/06/one-hot-encode-dna-sequence-using.html", target="none">One hot encoding DNA sequences using python.</a> ;</p>
	  </div>
    </div>
  </section>
    <section class="panel panel-default">
    <div class="panel-heading">
      <div class="panel-title">Archive</div>
    </div>
    <div class="list-group">
      <div class="list-group-item">
        <button class="accordion list-group-item">2018</button>
        <div class="accordian-panel panel">
         <p><a href="http://2-bitbio.com/2018/06/one-hot-encode-dna-sequence-using.html", target="none">One hot encoding DNA sequences using python.</a> ;</p>
          <p><a href="http://2-bitbio.com/2018/01/making-interactive-plots-using.html", target="none">Making interactive plots using javascript and D3 (v4)</a> ;</p>
        </div>
      </div>
    </div>
    <div class="list-group">
      <div class="list-group-item">
        <button class="accordion list-group-item">2017</button>
        <div class="accordian-panel">
          <p><a href="http://2-bitbio.com/2017/12/clickable-volcano-plots-in-shiny.html", target="none">Clickable Volcano Plots in Shiny!</a> ;</p>
         <p><a href="http://2-bitbio.com/2017/12/xkcd-plots.html", target="none">Xkcd plots!</a> ;</p>
         <p><a href="http://2-bitbio.com/2017/10/clustering-rnaseq-data-using-k-means.html", target="none">Clustering RNAseq data using K-means: how many clusters?</a> ;</p>
         <p><a href="http://2-bitbio.com/2017/09/bulk-download-stock-data-from-yahoo.html", target="none">Bulk download stock data from Yahoo finance with R</a> ;</p>
         <p><a href="http://2-bitbio.com/2017/09/pokemods-educational-outreach-initiative.html", target="none">Pokemodels: An outreach initiative</a> ;</p>
         <p><a href="http://2-bitbio.com/2017/07/blastx-is-too-slow-heres-some.html", target="none">BLASTx is too slow! Here's some alternatives.</a> ;</p>
         <p><a href="http://2-bitbio.com/2017/06/ramazon-get-your-shiny-app-on-amazon.html", target="none">Ramazon: Deploying shiny apps on AWS</a> ;</p>
         <p><a href="http://2-bitbio.com/2017/06/running-blast-in-shiny-web-application.html", target="none">Shiny BLAST</a> ;</p>
         <p><a href="http://2-bitbio.com/2017/05/using-shiny-to-make-web-plotter-for.html", target="none">Using Shiny to make a web plotter for RNAseq data</a> ;</p>
         <p><a href="http://2-bitbio.com/2017/04/animated-3d-pca-plots-in-r.html", target="none">Animated 3D PCA plots</a> ;</p>
         <p><a href="http://2-bitbio.com/2017/04/bash-commands-for-manipulating-fasta.html", target="none">Bash commands for manipulating fasta files</a> ;</p>
          <p><a href="http://2-bitbio.com/2017/04/clustering-rnaseq-data-making-heatmaps.html", target="none">Clustering RNAseq data, making heatmaps, and tree cutting to identify gene modules.</a> ;</p>
          <p><a href="http://2-bitbio.com/2017/04/rin-value-how-low-is-too-low.html", target="none">RIN value, how low is too low?</a> ;</p>
          <p><a href="http://2-bitbio.com/2017/03/rna-seq-do-i-need-replicates.html", target="none">RNAseq, how many replicates?</a> ;</p>
        </div>
      </div>
</aside>


  </div>
</div>

      </div>
    </main>

    <footer class="l-footer">
      <div class="container">
        <p><span class="h-logo">&copy; 2-BitBio</span></p>
        <aside>
          <p>Powered by <a href="https://gohugo.io/">Hugo</a>.</p>
          <p>Theme: <a href="https://github.com/dim0627/hugo_theme_beg">Beg</a></p>
        </aside>
      </div>
    </footer>

    <script src="//code.jquery.com/jquery-3.1.1.min.js"></script>
    <script src="//maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>
    <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/8.4/highlight.min.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>
    <script>
      var acc = document.getElementsByClassName("accordion");
      var i;

      for (i = 0; i < acc.length; i++) {
       acc[i].addEventListener("click", function() {
          

          this.classList.toggle("active");

          
         var panel = this.nextElementSibling;
         if (panel.style.display === "block") {
           panel.style.display = "none";
          } else {
            panel.style.display = "block";
         }
        });
      } 
    </script>
  </body>
</html>

